File: templates/captcha-verify/final.js
/**
 * final.js — 验证码逆向交付物【单一入口】（load → solve → verify 三段链路）。
 *
 * 这是 provider-neutral 骨架。真实平台协议必须由本 case 的 src/adapter.js 提供；
 * 本文件不假设厂商、接口名、HTTP 方法、JSONP、字段名或加密方案。
 *
 * 双重角色：
 *   - 自验：   node final.js            → 完整走 load→solve→verify→业务接口，交叉验证 5 次
 *   - 库调用： const { solveCaptcha, verifyChain } = require('./result');  → 只取 API，不自动执行
 *
 * 含 require.main 守卫。硬编码纪律（红线）：不含浏览器自动化代码；challenge 每次重新 load，不复用。
 *
 * 使用方式：
 *   node final.js                       # 默认：完整链路发真实请求，交叉验证 5 次
 *   node final.js --verify 5            # 指定验证次数
 *   node final.js --sign-only           # 仅输出 verify 参数（w 等），不发真实请求
 *   node final.js --cookie "name=value" # 注入用户 cookie（业务接口需要登录态时）
 *
 * answer JSON 契约见 references/captcha/captcha-overview.md；
 * 成功基线/失败复盘见 scripts/check_success_baseline.js + check_verification_attempts.js。
 */

'use strict';

const fs = require('fs');
const path = require('path');

// ============================================================
// 依赖（由用户从 templates 复制到 result/src/ 后填充）
// ============================================================
// 请求客户端：从 templates/node-request/client.js 复制到 result/src/request/client.js
const { createRequestSession, CookieJar } = require('./src/request/client');
// 真实平台协议适配器：必须由本 case 根据 trace/抓包实现。
let adapter = null;
try { adapter = require('./src/adapter'); } catch (_) { adapter = null; }
// 答案求解器：本地 ddddocr 或打码平台适配器，需导出 solve(imageBytes, type, options) → answer JSON
let solver = null;
try { solver = require('./src/solver'); } catch (_) { solver = null; }

// ============================================================
// 配置（静态外置 config.json + 内置默认）
// ============================================================
function loadConfig() {
  let cfg = {};
  try {
    cfg = JSON.parse(fs.readFileSync(path.join(__dirname, 'config.json'), 'utf8'));
  } catch (_) {}
  return Object.assign({
    target: { page_url: '<本 case 目标页>' },
    captcha: {
      provider: '<由本 case 证据确定>',
      captcha_type: '<由本 case 证据确定>',
    },
    solver: {
      mode: '<case-defined>',
      platform: '',
      api_key: '',
    },
    verify_count: 5,
  }, cfg);
}

// ============================================================
// 三段链路：load → solve → verify
// ============================================================

/**
 * ① load/bootstrap 阶段：完全交给 case adapter。
 */
async function loadChallenge(session, config) {
  assertAdapter();
  return adapter.loadChallenge(session, config);
}

/**
 * ② solve 阶段：下载素材 → 本地求解/人工接管/打码 → answer JSON
 * 返回：{ captcha_type, provider, offset/points, source_image_size, ... }（见 answer JSON 契约）
 *
 * solver 求解路径（按优先级）：
 *   1. 本地开源：ddddocr slide_match / OpenCV 模板匹配（solver.mode='ddddocr'）
 *   2. 人工接管：ddddocr/OpenCV 失效时（如拼图块重着色、背景像素扰动），用 scripts/click_gap.py 点击缺口
 *      → Node 侧通过 child_process 调 Python click_gap.py，或预存坐标后命令行传入
 *   3. 打码平台：solver.mode='platform'，走 solver_request_template.py 生成请求
 */
async function solveCaptcha(session, config, loadResult) {
  assertAdapter();
  const assets = await adapter.resolveAssets(session, config, loadResult);

  if (!solver) throw new Error('未配置 solver，请实现 result/src/solver.js');
  const answer = await solver.solve(assets.primary, config.captcha.captcha_type, {
    ...assets,
    provider: config.captcha.provider,
  });
  return adapter.prepareAnswer(answer, { session, config, loadResult, assets });
}

/**
 * ③ verify 阶段：加密 answer+track → 提交 → 换取通过凭据
 * 返回：由本 case adapter 定义的已验证凭据对象。
 */
async function verifyChain(session, config, loadResult, answer) {
  const request = await adapter.buildVerifyRequest({ session, config, loadResult, answer });
  if (!request || typeof request !== 'object' || !request.method || !request.url) {
    throw new Error('adapter.buildVerifyRequest 必须返回 { method, url, opts }');
  }
  const response = await session.request(request.method, request.url, request.opts || {});
  return adapter.parseVerifyResponse(response, { session, config, loadResult, answer, request });
}

/**
 * ④ 业务接口消费凭据
 */
async function callBusinessApi(session, config, credential) {
  return adapter.consumeCredential(session, config, credential);
}

// ============================================================
// 主流程：完整链路 + 交叉验证
// ============================================================
async function runOnce(config, cookieStr) {
  assertAdapter();
  const session = await createRequestSession({
    headers: cookieStr ? { Cookie: cookieStr } : {},
  });
  const jar = new CookieJar();
  session.defaults({ jar });

  try {
    const loadResult = await loadChallenge(session, config);
    const answer = await solveCaptcha(session, config, loadResult);
    const credential = await verifyChain(session, config, loadResult, answer);
    const bizResult = await callBusinessApi(session, config, credential);
    return { answer, credential, bizResult };
  } finally {
    if (session.close) session.close();
  }
}

async function main() {
  const args = require('minimist')(process.argv.slice(2), {
    boolean: ['sign-only'],
    alias: { verify: 'n', cookie: 'c' },
    default: { verify: null },
  });
  const config = loadConfig();
  assertAdapter();
  const verifyCount = args.verify || config.verify_count || 5;

  console.log(`[captcha-verify] provider=${config.captcha.provider} type=${config.captcha.captcha_type} verify=${verifyCount}`);

  if (args['sign-only']) {
    const session = await createRequestSession();
    try {
      const loadResult = await loadChallenge(session, config);
      const answer = await solveCaptcha(session, config, loadResult);
      const verifyRequest = await adapter.buildVerifyRequest({ session, config, loadResult, answer });
      console.log(JSON.stringify({ load: loadResult, answer, verifyRequest }, null, 2));
    } finally {
      if (session.close) session.close();
    }
    return;
  }

  let success = 0;
  for (let i = 1; i <= verifyCount; i++) {
    try {
      const result = await runOnce(config, args.cookie);
      success++;
      console.log(`  [${i}/${verifyCount}] OK  biz=${JSON.stringify(result.bizResult).slice(0, 100)}`);
    } catch (e) {
      console.error(`  [${i}/${verifyCount}] FAIL  ${e.message}`);
    }
  }
  console.log(`[captcha-verify] 完成 ${success}/${verifyCount}`);
  // 要求全部成功才算通过（与 README ≥5 次交叉验证一致）
  process.exit(success === verifyCount ? 0 : 1);
}

if (require.main === module) {
  main().catch(e => { console.error(e); process.exit(1); });
}

module.exports = { loadChallenge, solveCaptcha, verifyChain, callBusinessApi, runOnce };

function assertAdapter() {
  const required = ['loadChallenge', 'resolveAssets', 'prepareAnswer', 'buildVerifyRequest', 'parseVerifyResponse', 'consumeCredential'];
  if (!adapter) throw new Error('缺少 result/src/adapter.js：请先根据本 case 的真实 trace/抓包实现平台适配器');
  const missing = required.filter((name) => typeof adapter[name] !== 'function');
  if (missing.length) throw new Error(`adapter 缺少方法：${missing.join(', ')}`);
}

File: templates/captcha-verify-py/final.py
"""
final.py — 验证码逆向交付物【单一入口】（load → solve → verify 三段链路，Python 版）。

与 captcha-verify/final.js（Node 版）的区别：
  - Node 版 solver.js 调 ddddocr 需跨语言桥接（child_process/HTTP）
  - Python 版 solver.py 直接 import ddddocr，答案层工具链（ddddocr/OpenCV/Whisper）原生可用
  - 两版 config.json 字段一致，可共用；answer JSON 契约一致

双重角色：
  - 自验：   python final.py            → 完整走 load→solve→verify→业务接口，交叉验证 5 次
  - 库调用： from final import solve_captcha, verify_chain  → 只取 API，不自动执行

含 __main__ 守卫。硬编码纪律（红线）：不含浏览器自动化代码；challenge 每次重新 load，不复用。

使用方式：
  python final.py                       # 默认：完整链路发真实请求，交叉验证 5 次
  python final.py --verify 5            # 指定验证次数
  python final.py --sign-only           # 仅输出 verify 参数（w 等），不发真实请求
  python final.py --cookie "name=value" # 注入用户 cookie（业务接口需要登录态时）

answer JSON 契约见 references/captcha/captcha-overview.md；
坐标来源判定见 references/captcha/gap-coordinate-source.md；
成功基线/失败复盘见 scripts/check_success_baseline.js + check_verification_attempts.js。
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
import random

# ============================================================
# 依赖（由用户从 templates 复制到 result/src/ 后填充）
# ============================================================
# 请求客户端：从 templates/python-request/client.py 复制到 result/src/request/client.py
try:
    from src.request.client import create_request_session
except ImportError:
    create_request_session = None

# 真实平台协议适配器：必须由本 case 根据 trace/抓包实现。
try:
    from src.adapter import (
        load_challenge as adapter_load_challenge,
        resolve_assets as adapter_resolve_assets,
        prepare_answer as adapter_prepare_answer,
        build_verify_request as adapter_build_verify_request,
        parse_verify_response as adapter_parse_verify_response,
        consume_credential as adapter_consume_credential,
    )
except ImportError:
    adapter_load_challenge = adapter_resolve_assets = adapter_prepare_answer = None
    adapter_build_verify_request = adapter_parse_verify_response = adapter_consume_credential = None

# 答案求解器：ddddocr / OpenCV / 打码平台适配器，需导出 solve(image_bytes, captcha_type, options) → answer dict
try:
    from src.solver import solve
except ImportError:
    solve = None

# ============================================================
# 配置（provider-neutral；真实接口属于 case adapter）
# ============================================================
def load_config() -> dict:
    cfg = {}
    try:
        here = os.path.dirname(os.path.abspath(__file__))
        with open(os.path.join(here, "config.json"), encoding="utf-8") as f:
            cfg = json.load(f)
    except (FileNotFoundError, json.JSONDecodeError):
        pass
    defaults = {
        "target": {"page_url": "<本 case 目标页>"},
        "captcha": {
            "provider": "<由本 case 证据确定>",
            "captcha_type": "<由本 case 证据确定>",
        },
        "solver": {
            "mode": "<case-defined>",
            "platform": "",
            "api_key": "",
        },
        "verify_count": 5,
    }
    # 浅合并顶层键，嵌套 dict 用 defaults 兜底
    for key, val in defaults.items():
        if key not in cfg:
            cfg[key] = val
        elif isinstance(val, dict) and isinstance(cfg[key], dict):
            for sub_key, sub_val in val.items():
                cfg[key].setdefault(sub_key, sub_val)
    return cfg


# ============================================================
# 三段链路：load → solve → verify
# ============================================================

def load_challenge(session: object, config: dict) -> dict:
    """① bootstrap/load 阶段：完全交给 case adapter。"""
    require_adapter()
    return adapter_load_challenge(session, config)


def solve_captcha(session: object, config: dict, load_result: dict) -> dict:
    """② solve 阶段：下载素材 → 本地求解/打码 → answer JSON。"""
    if solve is None:
        raise RuntimeError("未配置 solver，请实现 result/src/solver.py")
    require_adapter()

    assets = adapter_resolve_assets(session, config, load_result)

    answer = solve(assets["primary"], config["captcha"]["captcha_type"], {
        **assets,
        "provider": config["captcha"]["provider"],
    })
    return adapter_prepare_answer(answer, session=session, config=config,
                          load_result=load_result, assets=assets)


def verify_chain(session: object, config: dict, load_result: dict, answer: dict) -> dict:
    """③ verify 阶段：加密 answer+track → 提交 → 换取通过凭据。"""
    require_adapter()
    request = adapter_build_verify_request(session=session, config=config,
                                   load_result=load_result, answer=answer)
    if not isinstance(request, dict) or not request.get("method") or not request.get("url"):
        raise ValueError("adapter.build_verify_request 必须返回 {method, url, opts}")
    opts = request.get("opts") or {}
    res = session.request(request["method"], request["url"], **opts)
    return adapter_parse_verify_response(res, session=session, config=config,
                                 load_result=load_result, answer=answer,
                                 request=request)


def call_business_api(session: object, config: dict, credential: dict) -> dict:
    """④ 业务接口消费凭据，由 case adapter 决定请求和成功判定。"""
    require_adapter()
    return adapter_consume_credential(session, config, credential)


# ============================================================
# 主流程：完整链路 + 交叉验证
# ============================================================
def run_once(config: dict, cookie_str: str = "") -> dict:
    require_adapter()
    headers = {"Cookie": cookie_str} if cookie_str else {}
    session = create_request_session(headers=headers)
    try:
        load_result = load_challenge(session, config)
        answer = solve_captcha(session, config, load_result)
        credential = verify_chain(session, config, load_result, answer)
        biz_result = call_business_api(session, config, credential)
        return {"answer": answer, "credential": credential, "biz_result": biz_result}
    finally:
        session.close()


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(description="验证码逆向交付物自验入口（Python）")
    parser.add_argument("--verify", type=int, default=5, help="交叉验证次数（默认 5）")
    parser.add_argument("--sign-only", action="store_true", help="仅输出 verify 参数，不发真实请求")
    parser.add_argument("--cookie", default="", help="注入用户 cookie")
    args = parser.parse_args(argv)

    config = load_config()
    verify_count = args.verify or config.get("verify_count", 5)

    require_adapter()
    print(f"[captcha-verify-py] provider={config['captcha']['provider']} "
          f"type={config['captcha']['captcha_type']} verify={verify_count}")

    if args.sign_only:
        session = create_request_session()
        try:
            load_result = load_challenge(session, config)
            answer = solve_captcha(session, config, load_result)
            request = adapter_build_verify_request(session=session, config=config,
                                                   load_result=load_result, answer=answer)
            print(json.dumps({"load": load_result, "answer": answer, "verify_request": request},
                             ensure_ascii=False, indent=2))
        finally:
            session.close()
        return 0

    success = 0
    for i in range(verify_count):
        try:
            result = run_once(config, args.cookie)
            success += 1
            biz_str = json.dumps(result["biz_result"], ensure_ascii=False)[:100]
            print(f"  [{i + 1}/{verify_count}] OK  biz={biz_str}")
        except Exception as e:
            print(f"  [{i + 1}/{verify_count}] FAIL  {e}")
        if i < verify_count - 1:
            time.sleep(1.0 + random.random() * 2.0)

    print(f"[captcha-verify-py] 完成 {success}/{verify_count}")
    # 要求全部成功才算通过（与 README ≥5 次交叉验证一致）
    return 0 if success == verify_count else 1


def require_adapter() -> None:
    required = {
        "load_challenge": adapter_load_challenge,
        "resolve_assets": adapter_resolve_assets,
        "prepare_answer": adapter_prepare_answer,
        "build_verify_request": adapter_build_verify_request,
        "parse_verify_response": adapter_parse_verify_response,
        "consume_credential": adapter_consume_credential,
    }
    missing = [name for name, fn in required.items() if not callable(fn)]
    if missing:
        raise RuntimeError("缺少 result/src/adapter.py 或方法：" + ", ".join(missing))


if __name__ == "__main__":
    sys.exit(main())
